single_kernel_mongo/status.py

```
@@ -39,17 +68,93 @@ def to_waiting(self, message: str):
+                "PBM": None,
+            }
+        )
+
+    def prioritize_statuses(self, statuses: StatusesDict) -> StatusBase:
+        """Prioritizes the statuses."""
+        mongodb_status, shard_status, config_server_status, pbm_status = (
+            statuses["mongodb"],
+            statuses["shard"],
+            statuses["config-server"],
+            statuses["PBM"],
+        )
+        if not isinstance(mongodb_status, ActiveStatus):
+            return mongodb_status
+
+        if shard_status and not isinstance(shard_status, ActiveStatus):
+            return shard_status
+
+        if config_server_status and not isinstance(config_server_status, ActiveStatus):
+            return config_server_status
+
+        if pbm_status and not isinstance(pbm_status, ActiveStatus):
+            return pbm_status
+
+        # if all statuses are active report mongodb status over sharding status
+        # This is also always the case for mongos charms because the 3 final statuses are None.
+        return mongodb_status
+
+    def process_and_share_statuses(self) -> None:
+        """Retrieves statuses from processes inside charm and shares the highest priority status.
+
+        When a non-fatal error occurs while processing statuses, the error is processed and
+        returned as a statuses.
+        """
+        # retrieve statuses of different services running on Charmed MongoDB
+        deployment_mode = (
+            "replica set" if self.state.is_role(MongoDBRoles.REPLICATION) else "cluster"
+        )
+        waiting_status = None
+        try:
+            statuses = self.get_statuses()
+        except OperationFailure as e:
+            if e.code in [13, 18]:
+                waiting_status = f"Waiting to sync passwords across the {deployment_mode}"
+            elif e.code == 133:
+                waiting_status = f"Waiting to sync internal membership across the {deployment_mode}"
+            else:
+                raise
```

How is this one handled? I am not sure this method should raise an exception, should it?

---

it isn't handled (in the original too), the exception is raised since it means the charm is in an error state in which no events should proceed.